### PR TITLE
Remove memory-format workaround for Arm backend

### PR DESCRIPTION
**Pull request by oscarandersson8218**
Remove temporary fix for memory format introduced in https://github.com/pytorch/executorch/pull/2371. The dim-order of each node is annotated in a pass. Also some refactoring of arm_backend.py.

### CODE REVIEW

File: backends/arm/passes/permute_memory_pass.py [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+            else:
+                data = node.meta["val"].data
+
+            if len(data.shape) == 4:
+                dim_order = NHWC_Order
+                if self.is_weight_node_for_dw_conv(node):
+                    dim_order = HWCM_Order
+            else:
+                dim_order = tuple(range(len(data.shape)))
+            node.meta["tosa_dim_order"] = dim_order
```
Comment (digantdesai) [inline backends/arm/passes/permute_memory_pass.py:53]: Let me think, there are at least three ways of doing this. 

(1) the way you are doing it. i.e. in `preprocess()` insert a dim_order tuple, track it manually and transform things as we convert to TOSA.
:+1: Flexibility
:+1: Mostly transparent to the user, expect the compile_spec
:-1: You are expecting the input to the `call_delegate()` to be in NHWC format
:-1: Doesn't use pytorch infra 
:-1: `node.meta['tosa_dim_order'] != node.meta["val"].dim_order()` which may cause confusion


(2) Global switch, i.e. `gm = export(model.to(memory_format=torch.channels_last), torch.randn(1,2,3,4).to(memory_format=torch.channels_last)`. This way `gm` should have everything in nhwc dim_order.
:+1: Clean impl, uses pytorch infra
:-1: If something falls out of delegate, that must be run in NHWC format
:-1: Assumes input conversion to NHWC happens outside of the model. Might be OK if you are getting it already in NHWC.

(3) Runtime conversion: After export we insert the `to_dim_order(NHWC)` ops in the beginning/end (also on the weights/placeholders etc.) as we want. Consume these nodes as we see fit in the delegate.
:+1: Mix of (1) and (2), allows you to select where the input conversion to NHWC happens.
:+1: Makes it flexible in terms of which node runs in which format. That is if you leave the node, it will take NCHW and convert it at runtime on the CPU. If you partition and delete the node the input to the delegate and to the model (I think, I have to check, this is similar to deleting first q node) will be NHWC.
:+1: Still uses pytorch infra, i.e. no need to manually maintain `tosa_dim_order` unless for cases like DW conv (but I see that as a special constraint from the backend so..)
:-1: slightly harder to implement, particularly if you are sprinkling multiple nodes in the graph. But if you run this pass before partitioner then it might not be too difficult.
:-1: Potentially involves running a pass outside i.e. modifying the graph before `to_backend()`. As long as this is user visible it might be ok.
:-1: Exposes some footguns when multiple delegates are involved but should be the case here.

File: backends/arm/passes/arm_pass.py [hunk 2]
```diff
@@ -0,0 +1,20 @@
...
+# All rights reserved.
+#
+# This source code is licensed under the BSD-style license found in the
+# LICENSE file in the root directory of this source tree.
+
+from executorch.exir.pass_base import ExportPass
+from torch.export import ExportedProgram
+
+
+class ArmPass(ExportPass):
```
[the remark this answers is not in the document]
Comment (oscarandersson8218) [reply backends/arm/passes/arm_pass.py:12]: For this PR, there is not rationale really. It was copied from https://github.com/pytorch/executorch/pull/3749. The rationale for introducing an ArmPass in that PR is to be able to pass an ExportedProgram to the pass to access weights and biases.

File: backends/arm/passes/permute_memory_pass.py (hunk 1 above)
Comment (oscarandersson8218) [reply backends/arm/passes/permute_memory_pass.py:53]: Thank you Digant for your comments, much appreciated!

(1) This would assume that the input is in NHWC-format if `permute_memory_to_nhwc==True`, yes. The next step of this PR could be to add a pass or extend this pass to add to_dim_order-ops around all our delegates such that the graph becomes: 
Input (default format) -> ... -> to_dim_order(nhwc) -> delegate -> to_dim_order(default format) -> ... -> output (default format). This would require that we modify the graph outside (just before and after) our delegate after partitioning. What's the proper way of doing this?
I agree with the last two bullets. Those are the drawbacks of the current solution.

(2) `model.to(memory_format=torch.channels_last)` and `node.meta["val"].to(memory_format=torch.channels_last)` gives `dim_order=(0,1,2,3)` for certain shapes such as (3,1,3,3). Otherwise I think `node.meta["val"].to(memory_format=torch.channels_last)` could have been used in (1) instead of annotating `node.meta["tosa_dim_order"]`.

(3) I don't think I fully understand this solution. We insert to_dim_order-ops at the inputs, outputs and placeholders of the full graph or just our delegate sub-graph? Will the dim_order be propagated forward when inserting to_dim_order-ops?

File: backends/arm/passes/arm_pass.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (oscarandersson8218) [reply backends/arm/passes/arm_pass.py]: Removed in latest commit.

**Comment by digantdesai** [comment]
I will take a look tomorrow.

File: backends/arm/passes/annotate_channels_last_dim_order_pass.py [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+            if node.target != dq_op:
+                return False
+            prev_node = node.args[0]
+            if prev_node.op != "placeholder":
+                return False
+            return is_consumer_node_depthwise_conv2d(node)
+        elif node.op == "placeholder":
+            # node is an input, weight or bias node
+            consumer_node = list(node.users)[0]
+            if self.is_weight_node_for_dw_conv(consumer_node):
```
Comment (digantdesai) [inline backends/arm/passes/annotate_channels_last_dim_order_pass.py:37]: nit: use consistent naming for depthwise convolution helper functions?

File: backends/arm/passes/annotate_channels_last_dim_order_pass.py [hunk 4]
```diff
@@ -0,0 +1,62 @@
...
+                node_data = node.meta["val"][0].data
+            else:
+                node_data = node.meta["val"].data
+
+            if len(node_data.shape) == 4:
+                dim_order = NHWC_Order
+                if self.is_weight_node_for_dw_conv(node):
+                    dim_order = HWCM_Order
+            else:
+                dim_order = tuple(range(node_data.dim()))
```
Comment (digantdesai) [inline backends/arm/passes/annotate_channels_last_dim_order_pass.py:59]: I assume this is mainly for the dim_order agnostic ops?

File: backends/arm/test/ops/test_linear.py [hunk 5]
```diff
@@ -120,7 +120,7 @@ def _test_linear_tosa_MI_pipeline(
             ArmTester(
                 module,
                 example_inputs=test_data,
-                compile_spec=common.get_tosa_compile_spec(),
+                compile_spec=common.get_tosa_compile_spec(permute_memory_to_nhwc=False),
```
Comment (digantdesai) [inline backends/arm/test/ops/test_linear.py:123]: I guess rank==4 ==> NHWC breaks this?

File: backends/arm/operators/op_placeholder.py [hunk 6]
```diff
@@ -7,167 +7,139 @@
...
+    input_shape = inputs[0].shape
+    input_dim_order = inputs[0].dim_order
+    tensor = ts.TosaSerializerTensor(
+        inputs[0].name,
+        tosa_shape(input_shape, input_dim_order),
+        ts.DType.INT8 if is_quant_arg(node) else inputs[0].dtype,
+        data=None,
+        placeholderFilename=inputs[0].name + ".npy",
+    )
+    tosa_graph.addInputTensor(tensor)
```
Comment (digantdesai) [inline backends/arm/operators/op_placeholder.py:35]: IIUC this means we will expect NHWC input tensor, do you want to add a runtime assert for this else it might just consume wrong format data and tricky to debug.

File: backends/arm/operators/op_add.py [hunk 7]
```diff
@@ -16,7 +16,7 @@
     build_rescale_from_int32,
     build_rescale_to_int32,
 )
-from executorch.backends.arm.tosa_utils import broadcast_shapes, getNodeArgs
+from executorch.backends.arm.tosa_utils import broadcast_shapes, getNodeArgs, tosa_shape
```
Comment (digantdesai) [inline backends/arm/operators/op_add.py:19]: Are you planning to add more ops? For instance Permute op argument needs to be updated based on what format input it gets when you update the previous node's output format.

**Review by digantdesai — APPROVED** [review]

File: backends/arm/passes/annotate_channels_last_dim_order_pass.py [hunk 8]
```diff
@@ -0,0 +1,62 @@
...
+                return True
+            if is_consumer_node_depthwise_conv2d(node):
+                # Check that node is the weight-argument and not input or bias
+                return consumer_node.args[1] == node
+
+        return False
+
+    def call(self, graph_module: torch.fx.GraphModule):
+        NHWC_Order = (0, 2, 3, 1)
+        HWCM_Order = (2, 3, 0, 1)
```
Comment (kimishpatel) [inline backends/arm/passes/annotate_channels_last_dim_order_pass.py:47]: I dont think I follow this dim order. Can you add some comments? Did you mean HWNC?

Comment (digantdesai) [reply backends/arm/passes/annotate_channels_last_dim_order_pass.py:47]: https://www.mlplatform.org/tosa/tosa_spec.html#_data_layouts

**Review by kimishpatel — COMMENTED** [review]
So not clear to me if this is really making use of dim order utils? is this in preparation to that?

**Comment by digantdesai** [comment]
> So not clear to me if this is really making use of dim order utils? is this in preparation to that?

Yes this is the intermediate step. See convo on Slack - https://pytorch.slack.com/archives/C01FV3A914N/p1719838104994989?thread_ts=1718355881.095489&cid=C01FV3A914N

File: backends/arm/passes/annotate_channels_last_dim_order_pass.py (hunk 8 above)
Comment (oscarandersson8218) [reply backends/arm/passes/annotate_channels_last_dim_order_pass.py:47]: It's a special case for depthwise convolution where the weights have shape = (H, W, C, M) => different dim-order (https://www.mlplatform.org/tosa/tosa_spec.html#_depthwise_conv2d).

File: backends/arm/test/ops/test_linear.py (hunk 5 above)
Comment (oscarandersson8218) [reply backends/arm/test/ops/test_linear.py:123]: Yes, (1, ,1, 1, C) works though. As a step towards to solution (2) we plan on removing this flag in the near future.

File: backends/arm/passes/annotate_channels_last_dim_order_pass.py (hunk 4 above)
Comment (oscarandersson8218) [reply backends/arm/passes/annotate_channels_last_dim_order_pass.py:59]: I consider it be for all ops, as we assume that the input is always transposed. It corresponds to this block https://github.com/pytorch/executorch/blob/main/backends/arm/arm_backend.py#L261C1-L261C70.

File: backends/arm/operators/op_add.py (hunk 7 above)
Comment (oscarandersson8218) [reply backends/arm/operators/op_add.py:19]: Yes, there are a few open PR:s with new ops. The ambition is that we can just update shapes such as permute's argument with `tosa_shape(shape, dim_order)`.

File: backends/arm/passes/annotate_channels_last_dim_order_pass.py (hunk 3 above)
Comment (oscarandersson8218) [reply backends/arm/passes/annotate_channels_last_dim_order_pass.py]: Updated in last rebase.

**Comment by digantdesai** [comment]
Rebase please?
